### PR TITLE
Spaces: Add sequential workspace switching and window move using sliding wall animation

### PR DESCRIPTION
**Pull request by goodwillcoding**
* update Tiling.selectSpace to accept optional use_mru argument which when
  false switches (and moves windows between) spaces sequencially (1->2->3)
* add keybinding for switch and move sequentially, using Super+PageUp/SuperDown
  as well as Super+Shift+PageUp/PageDown and removing those from second set of
  keybinding on switching/moving next/previous active workspace
* add prefs and dconf schemas for additional

Some notes: I avoided using switchWorkspaces leaving exclusively to non PaperWM uses as it stands now.

Please review and offer feedback. I intend to document and comment function some more as well, so do not pull PR yet (@hedning @olejorgenb)

### CODE REVIEW

**Comment by hedning** [comment]
:+1: I'd like to support this feature, but I think it needs its own animation. The carousel simply doesn't make much sense when navigating sequentially.

**Comment by goodwillcoding** [comment]
What are you thinking for animation? Showing slightly smaller clones but have them slide? Add arrows?

My original thinking was
1. Both next/prev and sequenctial are  workspace switching
2. Reluctance to introduce more code then needed
3. Gnome shell standard switcher I think is mostly unusable cause no previews (one used by switchWorkspaces)

That said in my work flow I rarely cycle thru all of them using Super+Above_Tab mru style so I might not be the best example since I mentally do not diffentiate between 2 types of switches but others might.

**Comment by hedning** [comment]
> Gnome shell standard switcher I think is mostly unusable cause no previews

Yeah, it also won't integrate with the per-workspace background. And since we keep a full clone of all the workspaces contained in a single actor it's pretty straightforward to make a another animation. We also don't want to trigger focus until navigation is done, which is again difficult to achieve with the native switcher.

> Showing slightly smaller clones but have them slide

Probably something like that. In general I think navigating through a fixed order vs. navigating through an mru . And in case of fixed order there's a pretty obvious way to visualize it as you note. Making `page_up/page_down` navigate sequentially seems like a good default.

Now, I think multimonitor needs some thought. Having one order which is shared between monitors isn't trivial (eg. workspaces that are visible on one monitor will leave gaps). The first solution that comes to mind is limiting sequential navigation to workspaces that are attached to the current monitor:
https://github.com/paperwm/PaperWM/blob/393365ed05e71194bdbdd2595486334960ff58cb/tiling.js#L946

File: tiling.js [hunk 1]
```diff
@@ -1476,19 +1483,46 @@ class Spaces extends Map {
...
+
+        // get current workspace and current workspaceindex
+        let currentSpace = this.spaceOf(workspaceManager.get_active_workspace());
+        let currentSpace_index = currentSpace.workspace.index();
+
+        // use the mru stack to get array of spaces
+        if (use_mru === true) {
+            spaces = [...this.stack];
+            // go over each monitor and collect spaces
+            this.monitors.forEach(space => spaces.splice(spaces.indexOf(space), 1));
```
Comment (hedning) [inline tiling.js]: Right, this line isn't very obvious :laughing: It goes through the spaces that are visible on a monitor and removes them from the space array as we don't want to duplicate them.

File: tiling.js [hunk 2]
```diff
@@ -1476,19 +1483,46 @@ class Spaces extends Map {
...
+            spaces = [currentSpace, ...spaces];
+        }
+        // get a sequential array of spaces
+        else {
+            let workspaces_count = workspaceManager.get_n_workspaces();
+            for (let i = 0; i < workspaces_count; i++) {
+                let workspace = workspaceManager.get_workspace_by_index(i);
+                spaces.push(this.get(workspace));
```
Comment (hedning) [inline tiling.js]: Here we might want to do something like this to limit the spaces to the ones that are on the current monitor:
```javascript
let space = this.get(workspace);
if (space.monitor === currentSpace.monitor)
  spaces.push(space)
```

**Comment by goodwillcoding** [comment]
@hedning okie everything you said and the code review comments make sense to me. I am going to do the monitor changes and work on slide style animation (might take some time, never done it before).

As a side note after your explanation I now understand why animation is a stack now and can not unsee it anyore ... and sequential space navigation using a stack does feel odd to me now ... lol.

**Comment by hedning** [comment]
> why animation is a stack

Hehe, yeah, I basically stole the idea from android's most recently used window stack.

Some stuff that might be useful:

- Pulling the sequential stuff into it's own method (eg. `switchSpace`). It then wouldn't hurt to rename `selectSpace` to something more sensible like `cycleSpace`, though that can be done after.
- We rely on the `window-manager::switch-workspace` signal to finish up animations via `animateToSpace`, so we'll need some sort of state which indicates what mode we were in. Using something like this `inPreview = {mode: Modes.MRU | Modes.SEQ}` might be the simplest (similar to how we handle `inGrab`). Again, it might be best to just add another method to handle finishing sequential workspace animation. 
- When navigating we probably want to prevent `super-above_tab` to work if we started navigating with `page_up/down` and vice versa.
- A space is contained within the monitor aligned `space.clip` actor, which you hopefully shouldn't need to touch as we're filtering out spaces that aren't already on the current monitor. So animations and scale should be applied to `space.actor` which takes monitor relative coordinates. You might need to add some padding between the `space.actor`s as the visible background is a bit larger than the monitor/`space.actor`.
- If you're an emacs user I'd recommend checking out https://github.com/paperwm/gnome-shell-mode, and if you're not it's still possible to use https://github.com/paperwm/gnome-shell-mode/blob/master/local/gnome-shell-mode/session.sh to launch nested gnome-shell sessions like this `session.sh x11 /path/to/paperwm/ paperwm@hedning:matrix.org` to test out the code.

Happy to answer any questions you might have :)

**Comment by goodwillcoding** [comment]
@hedning ... sorry it took so long, I was learning gnome shell extension development. I reimplemented the sequential navigation to your feedback and rebased my branch. When you get a chance can you please check it out.

All your feedback was implemented as far as I can tell, including restriction to monitor, but I am still testing multimonitor setup (I guess I need to test with dynamic space? [I usually use static ones]? right?)

Couple outstanding things.
1. I am not really liking InPreviewMode but ecmascript is not supporting enums, so meh
2. I am going to look into tweak clone placement so upper workspace is partially visible, assuming this approach work
3. Super-Above did not seem to be an issue, it actually works well between switcing sequential and stack modes nicely

**Comment by hedning** [comment]
No worries, the code base isn't really that pretty so big props to getting this done :)

Review incoming.

File: tiling.js [hunk 3]
```diff
@@ -1485,8 +1486,14 @@ class Spaces extends Map {
...
         let toSpace = this.spaceOf(to);
 
-        if (!inPreview)
-            this._initWorkspaceStack();
+        if (!inPreview) {
+            if (inPreviewMode.stack) {
+                this._initWorkspaceStack();
+            }
+            else if (inPreviewMode.sequence) {
+                this._initWorkspaceSequence();
```
Comment (hedning) [inline tiling.js]: I think it makes sense to default to the sequential animation.

File: tiling.js [hunk 4]
```diff
@@ -50,6 +50,7 @@ var sizeSlack = 30;
 var panelBox = Main.layoutManager.panelBox;
 
 var inPreview = false;
+var inPreviewMode = {stack: false, sequence: false};
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: completed

File: tiling.js [hunk 5]
```diff
@@ -1515,13 +1522,164 @@ class Spaces extends Map {
...
+        TopBar.updateWorkspaceIndicator(newSpace.workspace.index());
+
+        const scale = 0.8;
+        const padding = 25;
+        monitorSpaces.forEach((space, i) => {
+            let onComplete = () => {};
+            let actor = space.actor;
+
+            let space_height = ((space.height + padding) * (i - to) + padding) * scale;
+            space.actor.set_position(0, space_height);
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: implemented your code from  037f81e

File: tiling.js [hunk 6]
```diff
@@ -1515,13 +1522,164 @@ class Spaces extends Map {
...
+        if (Main.panel.statusArea.appMenu) {
+            Main.panel.statusArea.appMenu.container.hide();
+        }
+
+        let currentMonitor = currentSpace.monitor;
+        this.selectedSpace = currentSpace;
+
+        let cloneParent = currentSpace.clip.get_parent();
+
+        const scale = 0.8;
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: should that also be true in _initWorkspaceStack

File: tiling.js [hunk 7]
```diff
@@ -1515,13 +1522,164 @@ class Spaces extends Map {
...
+            let scaleY = currentMonitor.height / space.height;
+            space.clip.set_scale(scaleX, scaleY);
+
+            Tweener.removeTweens(space.border);
+            space.border.opacity = 255;
+            space.border.show();
+
+            space.actor.show();
+
+            let space_height = ((space.height + padding) * i + padding) * scale;
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: implemented from your commit

File: tiling.js [hunk 8]
```diff
@@ -1515,13 +1522,164 @@ class Spaces extends Map {
...
+
+        TopBar.updateWorkspaceIndicator(newSpace.workspace.index());
+
+        const scale = 0.8;
+        const padding = 25;
+        monitorSpaces.forEach((space, i) => {
+            let onComplete = () => {};
+            let actor = space.actor;
+
+            let space_height = ((space.height + padding) * (i - to) + padding) * scale;
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: makes sense

File: tiling.js [hunk 9]
```diff
@@ -1515,13 +1522,164 @@ class Spaces extends Map {
                 continue;
             monitor.clickOverlay.activate();
         }
+
+        inPreviewMode = {stack: false, sequence: false};
+    }
+
+    _initWorkspaceSequence() {
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: not sure what this one is, I do see animation completion in your commit is that what you meant

File: tiling.js [hunk 10]
```diff
@@ -1515,13 +1522,164 @@ class Spaces extends Map {
...
+            return;
+        }
+
+        // Always show the topbar when using the workspace stack
+        inPreview = true;
+        inPreviewMode = {stack: false, sequence: true};
+
+        TopBar.show();
+
+        let currentSpace = this.spaceOf(workspaceManager.get_active_workspace());
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: done

File: tiling.js [hunk 11]
```diff
@@ -1515,13 +1522,164 @@ class Spaces extends Map {
...
+            // Remove any lingering onComplete handlers from animateToSpace
+            Tweener.removeTweens(space.actor);
+
+            if (monitorSpaces[i - 1] === undefined) {
+                return;
+            }
+
+            let child = space.clip;
+            let sibling = monitorSpaces[i - 1].clip;
+            child !== sibling && cloneParent.set_child_below_sibling(child, sibling);
```
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: remove

File: tiling.js (hunk 3 above)
Comment (goodwillcoding) [reply tiling.js]: done

File: tiling.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (hedning) [reply tiling.js]: Yeah, I made quick hack that calls out to `_initWorkspaceSequence` in `animateToSpace`, but that requires resetting `inPreview` so not very pretty. Meaning we probably just want a shared method that sets up spaces in sequential order which both `_initWorkspaceSequence` and `animateToSpace` can use. Does that make sense?

File: tiling.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (hedning) [reply tiling.js]: That would probably make the stack animation far too busy, as we'd see all the spaces being scaled down at the same time.

[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: let me rephrase. _initWorkspaceSequence has a `scale=1` in this last revision but _initWorkspaceStack has scale of `0.9` ... so what I am wondering if _initWorkspaceStack scale should match _initWorkspaceSequence. (sorry if I am confused)

File: tiling.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: got it, will work on that

**Comment by goodwillcoding** [comment]
@hedning when you get a chance can you please check out 2e28be0. I improved the sequential animation positioning to show a small part of either next, previous or bothe next and previous worskpace which is a tad more intuitive and matches other switchers I have seen. There is a small cost of dropping scale from 0.9 to 0.825 (and for padding I use 4% of workspace height) which I also applied to the stack animation to match since it's possible to start with one and switch to another.

See it in action
Peek 2019-11-07 23-29

File: tiling.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (hedning) [reply tiling.js]: So `_initWorkspaceStack` (and `_initWorkspaceSequence`) is there to to set up the visual state before animations takes place, ie. position and potentially scale the non-visible spaces. So they serve the same function, but that doesn't mean they necessarily will share the same code.

For the workspace stack that means we want to position and scale the hidden spaces when the active space covers the the whole screen.

For sequential navigation we want to layout the hidden spaces in order «around» the active space. In this case we don't want the spaces to overlap at any time, partly because it looks better, but also to avoid having to manually manage the stacking (z-axis) order of the spaces. The easiest way to do this as far as I can see is having the spaces start out at `scale=1` and then scale down the sequence of spaces together.

So I'd say we should just leave the workspace stack alone for now. We might want to try to match the «active» scaling of the stack and sequence for visual consistency, but we can tackle that later (eg. the stack scaling is eg. tuned to always have overlap).

**Comment by hedning** [comment]
> I also applied to the stack animation to match since it's possible to start with one and switch to another.

Might be better to make the mode «stuck», so redirect any `select` to `switch` if we're in `sequential`. Not entirely sure, but going between sequential and stack doesn't seem very intuitive or useful to me. (If we land on them being used together then `select` and `switch` needs to set `inPreview`, otherwise the finishing animation won't work correctl (ie. `animateToSpace` won't know what to do).

File: tiling.js [hunk 12]
```diff
@@ -1665,7 +1836,7 @@ class Spaces extends Map {
     }
 
     animateToSpace(to, from, callback) {
-        inPreview = false;
+        inPreview = PreviewMode.NONE;
```
Comment (hedning) [inline tiling.js]: We'll need to remember the state here, otherwise we won't be able to select the correct finishing animation.

Comment (goodwillcoding) [reply tiling.js]: fixed.

**Comment by goodwillcoding** [comment]
> > I also applied to the stack animation to match since it's possible to start with one and switch to another.
> 
> Might be better to make the mode «stuck», so redirect any `select` to `switch` if we're in `sequential`. Not entirely sure, but going between sequential and stack doesn't seem very intuitive or useful to me. (If we land on them being used together then `select` and `switch` needs to set `inPreview`, otherwise the finishing animation won't work correctl (ie. `animateToSpace` won't know what to do).

I reverted the scale = 0.9 back to scale = 1. 
As a note, I h

File: tiling.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: So I looked over animateToSpace and switchToWorkspace and from what I see _initWorkspaceSequence is called before animateToSpace occurs and from what I can tell it's sufficient and have not had any side effects I can see after testing. I did put a conditional around some code in animateToSpace specific to stack arrangement but that was minor. please see 903dea3

also I removed a tween from _initWorkspaceSequence it seemed unneed to me (I could be wrong) but _initWorkspaceSequence also does not have it. Not sure what use case you were thinking of there, but in my testing it made no difference. See cfef02e

File: tiling.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (goodwillcoding) [reply tiling.js]: So I looked over animateToSpace and switchToWorkspace and from what I see _initWorkspaceSequence is called before animateToSpace occurs and from what I can tell it's sufficient and have not had any side effects I can see after testing. I did put a conditional around some code in animateToSpace specific to stack arrangement but that was minor. please see 903dea3

also I removed a tween from _initWorkspaceSequence it seemed unneed to me (I could be wrong) but _initWorkspaceSequence also does not have it. Not sure what use case you were thinking of there, but in my testing it made no difference. See cfef02e

**Comment by goodwillcoding** [comment]
> > I also applied to the stack animation to match since it's possible to start with one and switch to another.
> 
> Might be better to make the mode «stuck», so redirect any `select` to `switch` if we're in `sequential`. Not entirely sure, but going between sequential and stack doesn't seem very intuitive or useful to me. (If we land on them being used together then `select` and `switch` needs to set `inPreview`, otherwise the finishing animation won't work correctl (ie. `animateToSpace` won't know what to do).

Fixed in 0790409, now it's no longer possible to switch to stack when sequetial mode is on and vice versus.

**Comment by goodwillcoding** [comment]
@hedning renamed the navigation methods close to what you suggested -> d18e1d3

File: tiling.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (hedning) [reply tiling.js]: Currently `animateToSpace` only handles the selected space correctly when finishing a sequential animation. You can often see the space below overlapping the selected space.
image

This is fixable in two ways:
- Make sure the current space is raised above the other spaces
- Just animate all the whole sequence of spaces to its destination, this happens to involve the same code as `_initWorkspaceSequence`, but with animation. Meaning it makes sense to pull that out into its own method, and use it to both set up the sequence and finish the selection.

 Now I have some local fixes, so I'll push those to avoid any duplicate work.

**Comment by hedning** [comment]
Pushed a few fixes. In particular `animateToSpace` will now animate the whole sequence.

I'm happy with how things work/look now at least :) Just need to squash some commits and fixup commit messages, so they describe what's been done. Also keep the description to one line and not too long.

One remaining «issue» is animations being reset when releasing `Super`, though it's not a huge deal and will probably require quite a bit of refactoring (probably need to reparent the spaces into a single container that can be moved and scaled together). I'd say we don't bother with that for now, and rather land this first.

**Comment by goodwillcoding** [comment]
@hedning when you say squash **some** commits do you wantr to do all 21 commits both of us done into 1 or didi you mean you wanted to squash some of your more specific?

**Comment by goodwillcoding** [comment]
@hedning I have a 21 commit squashed on stand by with following message, let me know if it works for you
```
    Add sequential workspace switching and window move using sliding wall animation
    
    Move Super+PageUp/PageDown and Ctrl+Super+PageUp/PageDown from workspace stack
    to sequential workspace. Rename selectSpace to selectStackSpace. Rework
    animation meachanism to accomodate stack and sequential.
```

**Comment by hedning** [comment]
Took care of squashing/rebasing.

**Comment by goodwillcoding** [comment]
@hedning anything else you need before merging this? (I take it you do not want to squash all the commits into 1)

**Comment by goodwillcoding** [comment]
Also have this pending for the screenshot of sequential animation: https://github.com/paperwm/media/pull/1

**Comment by hedning** [comment]
Nope, think it's fine, will merge soon if I don't find any bugs etc. (should probably double check multiple monitors, only checked with that minimally).

**Comment by goodwillcoding** [comment]
@hedning works well on 3.28 multiple monitors. One monitor only has 1 space though (is that correct). I use 4 spaces on 2 dual monitors.

**Comment by hedning** [comment]
> @hedning works well on 3.28 multiple monitors. One monitor only has 1 space though (is that correct). I use 4 spaces on 2 dual monitors.

Yeah, since monitors can have different dimensions we don't want to move spaces without reason. Now that means we also want an easier way to move a space between monitors (eg. something like `examples.keybindings.moveSpaceToMonitor`).

**Comment by hedning** [comment]
We should probably show empty workspaces on all monitors. Will take a look in the future.
